### PR TITLE
feat(ngMock, ngMockE2E): add option to match latest definition for $h…

### LINKED ISSUE #11637: ngMock - why does $httpBackend match the first and not the last mock definition?
Is there any reason why `$httpBackend` matches the first and not the last mock definition? If it worked the other way around (i.e. `$httpBackend.when` added new mock definitions to the beginning of `definitions` array instead of the end of it) it would be much easier to overwrite already mocked requests.

I've got some default mock definitions common for almost all tests that are loaded before each test and would like to be able to easily overwrite them in some specific tests.

I wanted to prepare PR that changes this, but there's even a [unit test](https://github.com/angular/angular.js/blob/master/test/ngMock/angular-mocksSpec.js#L960) that explicitly checks for this behavior.

EDIT: In ngMock module unit tests like [`should match headers if specified`](https://github.com/angular/angular.js/blob/master/test/ngMock/angular-mocksSpec.js#L1011-L1013) or [`should match data if specified`](https://github.com/angular/angular.js/blob/master/test/ngMock/angular-mocksSpec.js#L1035-L1036), the first mocks are more specific than the later ones. In a real app, I'd expect something opposite - one would create general mocks for most tests and then define more specific ones in selected tests.

### PR DESCRIPTION
**Pull request by Narretz**
…ttpBackend request

Closes #16251
Closes #11637

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature

**What is the current behavior? (You can also link to an open issue here)**
See #11637 

**What is the new behavior (if this is a feature change)?**
Config option to change the behavior

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ?] Fix/Feature: Tests have been added; existing tests pass

**Other information**:
First draft to discuss the API.
As noted in #16251, it's not possible (afaics) to put the option on the httpBackendProvider, because they cannot be decorated. It's also not useful to put the method on the ng.httpBackendProvider, because that's not where it's needed. The option to switch it directly from httpBackend could also be used to migrate tests step by step or use it only for new tests. Another option would be a new "$httpBackendMocksProvider" or similar (but I don't think that's very useful).

### CODE REVIEW

File: src/ngMock/angular-mocks.js [hunk 1]
```diff
@@ -1430,8 +1431,9 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
       wasExpected = true;
     }
 
-    var i = -1, definition;
-    while ((definition = definitions[++i])) {
+    var i = matchLatestDefinition ? definitions.length : -1, definition;
+
+    while ((definition = definitions[matchLatestDefinition ? --i : ++i])) {
```
Comment (petebacondarwin) [inline src/ngMock/angular-mocks.js]: This is dangerous stuff due to the subtle difference between prefix and postfix operators. But given that we were doing this already I guess it'll be fine.

File: src/ngMock/angular-mocks.js [hunk 2]
```diff
@@ -1507,6 +1509,15 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
     return chain;
   };
 
+  $httpBackend.matchLatestDefinition = function(value) {
+
```
Comment (petebacondarwin) [inline src/ngMock/angular-mocks.js]: is this blank line needed / in keeping with the rest of the code?

File: src/ngMock/angular-mocks.js [hunk 3]
```diff
@@ -1507,6 +1509,15 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
...
+    if (isDefined(value)) {
+      matchLatestDefinition = value;
```
Comment (petebacondarwin) [inline src/ngMock/angular-mocks.js]: don't we normally return `this` when it is a setter to allow chaining?

File: test/ngMock/angular-mocksSpec.js [hunk 4]
```diff
@@ -1106,6 +1106,27 @@ describe('ngMock', function() {
...
+          expect(status).toBe(201);
+          expect(response).toBe('another');
+        });
+
+        hb('GET', '/url1', null, callback);
+        expect(callback).not.toHaveBeenCalled();
+        hb.flush();
+        expect(callback).toHaveBeenCalledOnce();
+      }
+    );
```
Comment (petebacondarwin) [inline test/ngMock/angular-mocksSpec.js]: Do we already have tests for the order when this is false?

**Review by gkalpak — APPROVED** [review]
LGTM except for the lack of docs 📖

File: test/ngMock/angular-mocksSpec.js [hunk 5]
```diff
@@ -1106,6 +1106,27 @@ describe('ngMock', function() {
...
+
+        callback.and.callFake(function(status, response) {
+          expect(status).toBe(201);
+          expect(response).toBe('another');
+        });
+
+        hb('GET', '/url1', null, callback);
+        expect(callback).not.toHaveBeenCalled();
+        hb.flush();
+        expect(callback).toHaveBeenCalledOnce();
```
Comment (petebacondarwin) [inline test/ngMock/angular-mocksSpec.js]: Should we check that it continues to work backwards if the last one doesn't match?

File: src/ngMock/angular-mocks.js [hunk 6]
```diff
@@ -1507,6 +1509,15 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
     return chain;
   };
 
+  $httpBackend.matchLatestDefinition = function(value) {
```
Comment (petebacondarwin) [inline src/ngMock/angular-mocks.js]: I think we needed docs here.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Looks good but needs docs and perhaps another test?

**Review by petebacondarwin — APPROVED** [review]

**Pull request closed without merging**